reconcile.go

```diff
@@ -151,77 +151,105 @@ func (a *Autopilot) applyDemotions(state *State, changes RaftChanges) (bool, err
 	return demoted, nil
 }
 
-// getFailedServers aggregates all of the information about servers that the consuming application believes are in
-// a failed/left state (indicated by the NodeStatus field on the Server type) as well as stale servers that are
-// in the raft configuration but not know to the consuming application. This function will do nothing with
-// that information and is purely to collect the data.
-func (a *Autopilot) getFailedServers() (*FailedServers, int, error) {
-	staleRaftServers := make(map[raft.ServerID]raft.Server)
-	raftConfig, err := a.getRaftConfiguration()
+func getRaftServerIds(servers []raft.Server) RaftServers {
+	ids := make(RaftServers)
+
+	for _, server := range servers {
+		ids[server.ID] = &VoterEligibility{
+			currentVoter: server.Suffrage == raft.Voter,
+		}
+	}
+
+	return ids
+}
+
+func (a *Autopilot) categorizeServers(voterPredicate func(NodeType) bool) (*CategorizedServers, error) {
+	cfg, err := a.getRaftConfiguration()
 	if err != nil {
-		return nil, 0, err
+		return nil, err
 	}
 
-	// Populate a map of all the raft servers. We will
-	// remove some later on from the map leaving us with
-	// just the stale servers.
-	var voters int
-	for _, server := range raftConfig.Servers {
-		staleRaftServers[server.ID] = server
+	// Get servers as raft sees them currently
+	// (we won't know if they have the potential to become voters yet)
+	raftServers := getRaftServerIds(cfg.Servers)
+	failedVoters := make(RaftServers)
+	failedNonVoters := make(RaftServers)
+	healthyVoters := make(RaftServers)
+	healthyNonVoters := make(RaftServers)
+
+	for id, srv := range a.delegate.KnownServers() {
+		v, found := raftServers[id]
+		if !found {
+			continue
+		}
+
+		delete(raftServers, id)
 
-		if server.Suffrage == raft.Voter {
-			voters++
+		if srv.NodeStatus == NodeAlive && v.IsCurrentVoter() {
+			healthyVoters[id] = v
+		} else if srv.NodeStatus == NodeAlive {
+			healthyNonVoters[id] = v
+		} else if v.IsCurrentVoter() {
+			failedVoters[id] = v
+		} else {
+			failedNonVoters[id] = v
 		}
+
+		v.SetPotentialVoter(voterPredicate(srv.NodeType))
 	}
 
-	var failed FailedServers
-	for id, srv := range a.delegate.KnownServers() {
-		raftSrv, found := staleRaftServers[id]
-		if found {
-			delete(staleRaftServers, id)
+	c := &CategorizedServers{
+		StaleNonVoters:   raftServers.FilterVoters(false),
+		StaleVoters:      raftServers.FilterVoters(true),
+		FailedNonVoters:  failedNonVoters,
+		FailedVoters:     failedVoters,
+		HealthyNonVoters: healthyNonVoters,
+		HealthyVoters:    healthyVoters,
+	}
+
+	return c, nil
+}
+
+func getFailureTolerance(nodes int) int {
+	return (nodes - 1) / 2
+}
+
+func isRemovalQuorate(voters int, minQuorum uint) bool {
+	return voters-1 >= int(minQuorum)
+}
+
+func removeStale(a *Autopilot) func(id raft.ServerID) error {
```

Comment:
Why not define these as methods on Autopilot and get rid of the func returns. 

You could pass a.removeStale to anything that would take a `func(raft.ServerID) error` argument.